### PR TITLE
feat: Add tf module

### PR DESCRIPTION
**Pull request by markbeierl**
# Description

Adds a plan for the user plane function as a machine charm.

# Checklist:

- [X] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
N/A I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
N/A Any dependent changes have been merged and published in downstream modules
N/A I have bumped the version of the library

### CODE REVIEW

File: terraform/README.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
+# SD-Core UPF Terraform Module
+
+This folder contains a base [Terraform][Terraform] module for the sdcore-gnbsim charm.
```
Comment (ghislainbourgeois) [inline terraform/README.md:3]:
```suggestion
This folder contains a base [Terraform][Terraform] module for the sdcore-upf machine charm.
```

File: terraform/README.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  the application name.
+- **terraform.tf** - Defines the Terraform provider.
+
+## Deploying sdcore-upf base module separately
+
+### Pre-requisites
+
+- A host with a CPU supporting AVX2 and RDRAND instructions (Intel Haswell, AMD Excavator or equivalent)
+- Juju controller bootstrapped to a LXD cluster
+- A Juju machine manually representing the host added to the controller
```
Comment (ghislainbourgeois) [inline terraform/README.md:28]:
```suggestion
- A Juju machine representing the host manually added to the controller
```

File: terraform/main.tf [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+resource "juju_application" "upf" {
+  name  = var.app_name
+  model = var.model_name
+
+  charm {
+    name    = "sdcore-upf"
+    channel = var.channel
+  }
+  config = var.config
+  trust  = true
```
Comment (ghislainbourgeois) [inline terraform/main.tf:13]: Does this do anything with a machine charm? I think we can drop it otherwise.

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]
Looks good, some small details to fix.

File: terraform/main.tf (hunk 3 above)
Comment (markbeierl) [reply terraform/main.tf]: You're right, good catch, thanks

**Review by ghislainbourgeois — APPROVED** [review]

**Review by gatici — APPROVED** [review]
Looks good to me. Provided a comment.

File: terraform/README.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+From inside the `terraform` folder, initialize the provider:
+
+```shell
+terraform init
+```
+
+Create Terraform plan:
+
+```shell
+terraform plan
```
[the remark this answers is not in the document]
Comment (markbeierl) [reply terraform/README.md:44]: L28 states that you need the machine already created, which you cannot do without a model. Is it worth mentioning it in the pre-requisites, or is implicit good enough?

[the remark this answers is not in the document]
Comment (gatici) [reply terraform/README.md:44]: The latest commit makes it clear enough. Thank you!

**Review by gatici — APPROVED** [review]

**Pull request merged**